### PR TITLE
Cfq bahdanau to luong decoder

### PR DESCRIPTION
**Pull request by Ranapop**
This PR changes the decoder from Bahdanau's decoder to Luong's decoder

In Bahdanau's decoder the **context vector** is computed using the **previous RNN** state as keys. The context vector is fed into the RNN together with the decoder inputs, and the context vector with the input and current RNN state are passed through a linear layer to compute the pre-output. The pre-output is then used to compute the final predictions (passed through another projection layer).

In Luong's decoder the context vector is computed using the **current state**, not the previous RNN state. Also, the context vector and current RNN state are used to create the so called **attention vector** (concatenation + linear layer). This attention vector is then used for the final predictions (projection layer), resulting in a simpler flow compared to Bahdanau decoder.  The attention vector computed at the previous time step is fed into the RNN at the current step, and a zero-valued vector is used for the first step (as done in [Joey decoder](https://joeynmt.readthedocs.io/en/stable/_modules/joeynmt/decoders.html) ).

This changes were made as in practice Luong's decoder is being used more often, and the tensor2tensor implementation also uses this type of decoder. 

The changes were tested together with other changes (dropout, gradient clipping) and led to a relatively smooth training curve on 400 epochs. The other changes will be merged in follow-up PRs.

### CODE REVIEW

File: examples/cfq/models.py [hunk 1]
```diff
@@ -189,8 +194,8 @@ def apply(self,
     """
     multilayer_lstm_cell = MultilayerLSTM.partial(num_layers=num_layers).shared(
         name='multilayer_lstm')
-    pre_output_layer = nn.Dense.shared(features=DECODER_PROJECTION,
-                                       name='pre_output_layer')
+    attention_layer = nn.Dense.shared(features=ATTENTION_LAYER_SIZE,
+                                       name='attention_layer')
```
Comment (marcvanzee) [inline examples/cfq/models.py]:
```suggestion
    attention_layer = nn.Dense.shared(features=ATTENTION_LAYER_SIZE,
                                      name='attention_layer')
```

File: examples/cfq/models.py [hunk 2]
```diff
@@ -212,32 +217,33 @@ def apply(self,
...
+      context = mlp_attention(jnp.expand_dims(h, 1), projected_keys,
+                              encoder_hidden_states, attention_mask)
+      context_and_state = jnp.concatenate([context, h], axis=-1)
+      attention = jnp.tanh(attention_layer(context_and_state))
+      logits = projection(attention)
       predicted_tokens = jax.random.categorical(categorical_rng, logits)
       predicted_tokens_uint8 = jnp.asarray(predicted_tokens, dtype=jnp.uint8)
-      return (carry_rng, (states, h),
-              predicted_tokens_uint8), (logits, predicted_tokens_uint8)
+      new_carry = (carry_rng, (states, h), predicted_tokens_uint8, attention)
```
Comment (marcvanzee) [inline examples/cfq/models.py]: I guess we don't have to pass `h` here, since we don't use it in the next step anyway.

**Review by marcvanzee — APPROVED** [review]
